ClickHouse support

sazonov opened the pull request:
Essential ClickHouse support.

sazonov commented:
@ DoodleBobBuffPants could you please review this PR?

sazonov commented:
@DoodleBobBuffPants could you please review this PR?

Hubbitus commented:
Are there some problems by merging Clickhouse support?

Hubbitus commented on flyway-community-db-support/pom.xml:
```diff
@@ -45,12 +45,22 @@
+            <groupId>com.clickhouse</groupId>
+            <artifactId>clickhouse-jdbc</artifactId>
+            <version>0.4.0</version>
+            <optional>true</optional>
```
Why dependency is optional and driver not included into build by default?

sazonov replied:
Because you may want to use a different version of driver or even legacy driver from "ru.yandex" package.

sazonov replied:
You can find that all drivers are optional in pom.xml for flyway-core and flyway-community-db-support modules. Your app has to include a proper driver.

Hubbitus commented on flyway-community-db-support/src/main/java/org/flywaydb/community/database/clickhouse/ClickHouseDatabase.java:
```diff
@@ -0,0 +1,111 @@
+                    "    execution_time Int32," +
+                    "    success Bool" +
+                    ")" +
+                    " ENGINE = StripeLog;" +
```
On the repair command got: `SQL Error [48] [07000]: Code: 48. DB::Exception: Table engine StripeLog doesn't support mutations. (NOT_IMPLEMENTED) (version 23.2.1.2537 (official build))`

I would like to suggest replacing `ENGINE = StripeLog;` by `ENGINE = MergeTree PRIMARY KEY `version`;`

sazonov replied:
Thanks, fixed that.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/flyway/flyway?pullRequest=3611) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/flyway/flyway?pullRequest=3611) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: sazonov<br/>:x: rg-buildmonkey<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/flyway/flyway?pullRequest=3611) it.</sub>

Hubbitus commented:
@sazonov, there appeared so many changes are not relevant to the change... Is it intentional? Could you please rebase pull request to the actual state of master branch?

sazonov commented:
> @sazonov, there appeared so many changes are not relevant to the change... Is it intentional? Could you please rebase pull request to the actual state of master branch?

Done. There were changes that actually exist in target branch. This was result of syncing forks via github "sync fork" button, seems like it's buggy.

Mityushin commented on flyway-community-db-support/src/main/java/org/flywaydb/community/database/clickhouse/ClickHouseConfigurationExtension.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Override
+    public void extractParametersFromConfiguration(Map<String, String> configuration) {
+        clusterName = configuration.get(CLUSTER_NAME);
```
```java
String clusterName = configuration.remove(CLUSTER_NAME);
if (clusterName != null) {
    this.clusterName = clusterName;
}
```
I think need to be like this. 
This solution was tested by me with spring boot flyway autoconfiguration. This method was calling multiple times during spring context initialization

sazonov replied:
Do you have several Flyway configurations in your app? Each flyway extension (plugin) expects to be configured only once. PostgreSQLConfigurationExtension has the same issue if it is reconfigured. I fixed that for ClickHouse, but it looks like some problem with app configuration.

Mityushin commented on flyway-community-db-support/src/main/java/org/flywaydb/community/database/clickhouse/ClickHouseDatabase.java:
```diff
@@ -0,0 +1,112 @@
+        String clusterName = configuration.getPluginRegister().getPlugin(ClickHouseConfigurationExtension.class).getClusterName();
+
+        if (StringUtils.hasText(clusterName)) {
+            return "CREATE TABLE IF NOT EXISTS " + table + "_local ON CLUSTER " + clusterName + "(" +
```
This concatination is wrong. Result string:
```
CREATE TABLE IF NOT EXISTS "metrics_db"."flyway_schema_history"_local ON CLUSTER test_shard_four_shards_localhost(    installed_rank Int32,    version String,    description String,    type String,    script String,    checksum Nullable(Int32),    installed_by String,    installed_on DateTime DEFAULT now(),    execution_time Int32,    success Bool) ENGINE = ReplicatedMergeTree(   '/clickhouse/tables/{shard}/flyway_schema_history',   '{replica}' ) PARTITION BY tuple() ORDER BY (installed_rank)
```
Server error:
```
Code: 62. DB::Exception: Syntax error: failed at position 64 ('_local'): _local ON CLUSTER test_shard_four_shards_localhost(    installed_rank Int32,    version String,    description String,    type String,    script String,    chec. Expected one of: token, Dot, UUID, ON, OpeningRoundBracket, storage definition, ENGINE, PARTITION BY, PRIMARY KEY, ORDER BY, SAMPLE BY, TTL, EMPTY AS, AS, COMMENT, INTO OUTFILE, FORMAT, SETTINGS, end of query. (SYNTAX_ERROR) (version 22.9.6.20 (official build))
, server ClickHouseNode [uri=http://localhost:49635/metrics_db]@206826844
```

Mityushin replied:
This error is repeatable on my branch (for now not access on github). I was port you PR to [flyway-8.5.13](https://github.com/flyway/flyway/releases/tag/flyway-8.5.13) because [spring-boot 2.7.x](https://github.com/spring-projects/spring-boot/blob/2.7.x/spring-boot-project/spring-boot-dependencies/build.gradle#L347) depend on this version of library.

Main difference 8.5.13 & 9.15.1 is how we access clusterName. In 8.5.13
```java
String clusterName = PluginRegister.getPlugin(ClickHouseConfigurationExtension.class).getClusterName();
```
Anyway if you have this bug in [flyway-9.15.1](https://github.com/flyway/flyway/releases/tag/flyway-9.15.1) this should help you
```java
String tableName = table.getName();
String localTableNameQuoted = quote(tableName + "_local");
```

sazonov replied:
I've reworked cluster table creation and will push it today. I recommend wait for the fix, because there will be no Distributed table because it doesn't support ClickHouse mutations and flyway "repair" command doesn't work.

sazonov replied:
> This error is repeatable on my branch (for now not access on github). I was port you PR to [flyway-8.5.13](https://github.com/flyway/flyway/releases/tag/flyway-8.5.13) because [spring-boot 2.7.x](https://github.com/spring-projects/spring-boot/blob/2.7.x/spring-boot-project/spring-boot-dependencies/build.gradle#L347) depend on this version of library.

I would suggest using "flyway.version" variable in your Spring Boot app to use actual flyway build and avoid backporting.

Mityushin replied:
Thank you. That works

Mityushin commented on flyway-community-db-support/src/main/java/org/flywaydb/community/database/clickhouse/ClickHouseTable.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    protected void doDrop() throws SQLException {
+        jdbcTemplate.executeStatement("DROP TABLE " + database.quote(schema.getName(), name));
```
Some cases require `ON CLUSTER <cluster_name>` suffix, from ClickHouseConfigurationExtension

sazonov replied:
Thanks. Fixed that.

Mityushin commented on flyway-community-db-support/pom.xml:
```diff
@@ -45,12 +45,22 @@
+        <dependency>
+            <groupId>com.clickhouse</groupId>
+            <artifactId>clickhouse-jdbc</artifactId>
+            <version>0.4.0</version>
```
0.4.1 available
https://github.com/ClickHouse/clickhouse-java/releases/tag/v0.4.1

sazonov replied:
Updated.

Mityushin commented:
@sazonov thank you very much for excellent work. Your input is very helpful, you helped us get into production. Thank you!

Mityushin commented:
@DoodleBobBuffPants LGTM

Mityushin commented:
@tjs145 please look at this PR

Hubbitus commented:
@sazonov, could you please look at https://github.com/flyway/flyway/issues/3633? Is it clickhouse related or should be handled in a more common way?

sazonov commented:
> @sazonov, could you please look at #3633? Is it clickhouse related or should be handled in a more common way?

It looks like it's not expected that anyone uses sensitive information in placeholders. All placeholders get to additional jdbc driver parameters. From here: org/flywaydb/core/api/configuration/ClassicConfiguration.java:1672 and eventually goes here: org/flywaydb/core/internal/jdbc/DriverDataSource.java:258. Placeholders can be overwritten on the DatabaseType level but there is no way to determine whether they are sensitive data or not.

Hubbitus commented:
@sazonov, could you please update your PR to newer version of flyway 9.16.0?

Today I got `ERROR: No database found to handle jdbc:clickhouse://10.220.1.10:8123/datamart`

9.15.2 works fine with Clickhouse, but was regression wit Postgres, when `system.tables` attempted to read.

sazonov commented on flyway-community-db-support/src/main/java/org/flywaydb/community/database/clickhouse/ClickHouseDatabaseType.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Override
+    public boolean handlesDatabaseProductNameAndVersion(String databaseProductName, String databaseProductVersion, Connection connection) {
+        return true;
```
Fixed.

Hubbitus commented on flyway-community-db-support/src/main/java/org/flywaydb/community/database/clickhouse/ClickHouseDatabase.java:
```diff
@@ -0,0 +1,103 @@
+
+        String script = "CREATE TABLE IF NOT EXISTS " + table + (isClustered ? (" ON CLUSTER " + clusterName) : "") + "(" +
+                "    installed_rank Int32," +
+                "    version String," +
```
Unfortunately, on repeatable migrations version will be null and that failed with:
```
ERROR: Failed to execute operation in non-transactional context. Please restore backups and roll back database and code!
ERROR: Unable to insert row for version 'null' in Schema History table "datamart"."__flyway_schema_history"
----------------------------------------------------------------------------------------------------
SQL State  : HY000
Error Code : 0
Message    : Cannot set null to non-nullable column #2 [version String]

Caused by: java.sql.SQLException: Cannot set null to non-nullable column #2 [version String]
```

`installed_rank` should be PRIMARY KEY and not-nullable, but version `Nullable(String)`.

So, resulting table should look like:
```sql
CREATE TABLE datamart.__flyway_schema_history_NEW ON CLUSTER ${CLUSTER_NAME} (
    `installed_rank` Int32,
    `version` Nullable(String),
    `description` String,
    `type` String,
    `script` String,
    `checksum` Nullable(Int32),
    `installed_by` String,
    `installed_on` DateTime DEFAULT now(),
    `execution_time` Int32,
    `success` Bool
)
ENGINE = MergeTree
PRIMARY KEY `installed_rank`
ORDER BY `installed_rank`
SETTINGS index_granularity = 8192;
```

sazonov replied:
I think it's better to set field 'script' as primary key. Field 'installed_rank has no connection with migration content.

sazonov replied:
@Hubbitus could you please check the fix.

Hubbitus replied:
For postgres variant `installed_rank` is the primary key.

sazonov replied:
Maybe they should change that for PG as well.

Hubbitus replied:
[Ignite as well](https://github.com/flyway/flyway/blob/08ff5b0c6ecf2a87e32e41d72904b7ec92a522de/flyway-community-db-support/src/main/java/org/flywaydb/community/database/ignite/thin/IgniteThinDatabase.java#L75), I also checked `mysql`, `sqlite` and some others. All has `installed_rank` as `PRIMARY KEY`.
Meantime, it looks like for the Clickhouse we may omit it completely? So, there is even no unique constraint will be implied...

I think it is not consistent to set it to the `script` value. But I'm uncertain.

sazonov replied:
Primary key in this table prevents duplications when applying migrations, but field 'installed_rank' contains auto incremented value and it's useless for this purpose. Field 'script' contains version number and migration name, it's unique and it's a perfect candidate for the primary key. Let's wait for maintainer to review this PR.

mshustov commented:
@sazonov @Hubbitus, what open questions are left to have the PR merged? Could you clarify what help is needed to pull the PR over the line?

sazonov commented:
> @sazonov @Hubbitus, what open questions are left to have the PR merged? Could you clarify what help is needed to pull the PR over the line?

No help is needed. PR is waiting to be merged by @Barry-RG. I'm syncing PR with master to keep it actual.

kostinkirill commented:
Hey guys, is there any chance to see it in the next release?

BOOMeranGG commented:
I'm waiting for this PR too

noorul commented:
Waiting for this PR to be merged and released.

thetimur commented:
We're waiting for this PR with our team

Hope it will be merged and released asap

bahusdivus commented:
It would be nice to see it released soon!

lemonl2 commented:
@sazonov if we need to create clickhouse distributed table and logic table，how to do it?

echastukhin commented:
@sazonov thank you very much for your work to add support of clickhouse.
Can you help with building of version 9.20.1 from clickhouse-support branch?
command 'mvn clean install -Pbuild-assemblies-linux' rise error 
"org.flywaydb.community.database.clickhouse.ClickHouseConfigurationExtension is not abstract and does not override abstract method getNamespace() in org.flywaydb.core.extensibility.ConfigurationExtension"

sazonov commented:
> @sazonov thank you very much for your work to add support of clickhouse. Can you help with building of version 9.20.1 from clickhouse-support branch? command 'mvn clean install -Pbuild-assemblies-linux' rise error "org.flywaydb.community.database.clickhouse.ClickHouseConfigurationExtension is not abstract and does not override abstract method getNamespace() in org.flywaydb.core.extensibility.ConfigurationExtension"

@echastukhin you are welcome. Fixed. I updated ClickHouse JDBC driver as well. Tested on single node and cluster env.

andrewald commented:
We are also waiting for this PR to finally use Flyway with Clickhouse. Would be nice if this gets released soon!

pluto-tu commented:
clickHouse is coming soon?

sazonov commented:
> > No help is needed. PR is waiting to be merged by @Barry-RG. I'm syncing PR with master to keep it actual.
> 
> The Flyway team is working on v10. The new DB support is on hold until the new major version is available (ETA ~ the upcoming fall). @sazonov Could you update the branch once v10 is available? I'll resume the talks with the Flyway team to merge the PR.

Okay, thanks.

bbeckerdarosa commented:
@sazonov @Barry-RG Do you have any estimate when this PR will be merged?
I believe due to a new release of flyway on September 21 this branch is out of date because flyway-parent is throwing some errors, but it may be that I'm doing something wrong trying to use the feature of this branch.

sazonov commented:
> @sazonov @Barry-RG Do you have any estimate when this PR will be merged? I believe due to a new release of flyway on September 21 this branch might be out of date because either flyway-parent or flyway-core are throwing some errors, but it may be that I'm doing something wrong trying to use the feature of this branch.

There were two minor releases, I synced them with the PR. Everything worked well before and still works.

mitchbregs commented:
Hey @sazonov - I am trying to wrap all of this in a single multistage Docker build. Here is what I have so far:

```
FROM maven as builder

COPY pom.xml /tmp/pom.xml
WORKDIR /tmp
RUN mvn dependency:copy-dependencies

RUN git clone -b clickhouse-support https://github.com/sazonov/flyway.git /flyway-source
WORKDIR /flyway-source
RUN mvn clean install

FROM flyway/flyway:8.0.2

COPY --from=builder /tmp/target/dependency/*.jar /flyway/drivers/
COPY --from=builder /flyway-source/flyway-community-db-support/target/flyway-community-db-support-*.jar /flyway/lib/community/
```

where `pom.xml` is: 

```
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>com.example</groupId>
    <artifactId>clickhouse-dependencies</artifactId>
    <version>1.0-SNAPSHOT</version>

    <dependencies>
        <dependency>
            <groupId>com.clickhouse</groupId>
            <artifactId>clickhouse-jdbc</artifactId>
            <version>0.4.6</version>
        </dependency>
    </dependencies>
</project>
```

When I try to run:
```
$ docker run flyway-clickhouse:latest -url=jdbc:clickhouse://... -user=... -password=... migrate
```

Error:
```
ERROR: No database found to handle jdbc:clickhouse://
```

Not a Java developer, so not entirely familiar with Maven. Any suggestions?

thgruiz commented:
so... tried you Dockerfile, Mitchell, and this is what I found:

1) I think this branch is ready to be used with flyway 9:

changing your:
FROM flyway/flyway:8.0.2
to
FROM flyway/flyway:9.21.2
it gives me a 'further error' of connection refused (bc i did not pointed
to a real instance of clickhouse :) )

docker run --rm
 ddb0e9f378c895f9538cad13d43a374637a651bfd41973138a7d259e42cc3854
 -url=jdbc:clickhouse://127.0.0.1:8123 -user=any -password=any migrate

WARNING: This version of Flyway is out of date. Upgrade to Flyway 9.22.2:
https://rd.gt/3rXiSlV

Flyway Community Edition 9.21.2 by Redgate
See release notes here: https://rd.gt/416ObMi

Flyway is performing some final checks. Thank you for your patience.
ERROR: Unable to obtain connection from database (jdbc:clickhouse://
127.0.0.1:8123) for user 'any': Connection refused, server ClickHouseNode
***@***.***
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL State  : 08000
Error Code : 210
Message    : Connection refused, server ClickHouseNode [uri=
***@***.***

Caused by: java.sql.SQLException: Connection refused, server ClickHouseNode
***@***.***
Caused by: java.net.ConnectException: Connection refused





2) In your image I ended up with only the clickhouse-jdbc driver:

docker run -it --rm --entrypoint /bin/bash
ddb0e9f378c895f9538cad13d43a374637a651bfd41973138a7d259e42cc3854  -c "ls
drivers | grep house"
clickhouse-jdbc-0.4.6.jar

if we use an older jdbc driver version all other needed jars are copied as
well:

(using clickhouse-jdbc v0.3.2 on pom.xml):

docker run -it --rm --entrypoint /bin/bash
effaab4792f7dbd71f9cf8be3bda3f60a2a1039499b39739b70109595520882c   -c "ls
drivers | grep house"
clickhouse-grpc-client-0.3.2-netty.jar
clickhouse-http-client-0.3.2-shaded.jar
clickhouse-jdbc-0.3.2.jar

So it seems that the new 0.4.6 jars don't have all dependencies in order.
You should add the other dependencies jars (the http-client one) in pom (or
use the older driver version) so you will not run on the 'no transport
available' (or something like that) further ahead



hope this helps,

Thiago Ruiz


Em sex., 6 de out. de 2023 às 12:28, Mitchell Bregman <
***@***.***> escreveu:

> Hey @sazonov <https://github.com/sazonov> - I am trying to wrap all of
> this in a single multistage Docker build. Here is what I have so far:
>
> FROM maven as builder
>
> COPY pom.xml /tmp/pom.xml
> WORKDIR /tmp
> RUN mvn dependency:copy-dependencies
>
> RUN git clone -b clickhouse-support https://github.com/sazonov/flyway.git /flyway-source
> WORKDIR /flyway-source
> RUN mvn clean install
>
> FROM flyway/flyway:8.0.2
>
> COPY --from=builder /tmp/target/dependency/*.jar /flyway/drivers/
> COPY --from=builder /flyway-source/flyway-community-db-support/target/flyway-community-db-support-*.jar /flyway/lib/community/
>
> where pom.xml is:
>
> <project xmlns="http://maven.apache.org/POM/4.0.0"
>          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
>          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
>     <modelVersion>4.0.0</modelVersion>
>
>     <groupId>com.example</groupId>
>     <artifactId>clickhouse-dependencies</artifactId>
>     <version>1.0-SNAPSHOT</version>
>
>     <dependencies>
>         <dependency>
>             <groupId>com.clickhouse</groupId>
>             <artifactId>clickhouse-jdbc</artifactId>
>             <version>0.4.6</version>
>         </dependency>
>     </dependencies>
> </project>
>
> When I try to run:
>
> $ docker run flyway-clickhouse:latest -url=jdbc:clickhouse://... -user=... -password=... migrate
>
> Error:
>
> ERROR: No database found to handle jdbc:clickhouse://
>
> Not a Java developer, so not entirely familiar with Maven. Any suggestions?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/flyway/flyway/pull/3611#issuecomment-1750909362>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI56LQCRG32J2LWPEGMJCTDX6APR7AVCNFSM6AAAAAAUQDXGE2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONJQHEYDSMZWGI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Thiago Ruiz.

nyivanov commented:
@sazonov Can you please post an example on how to use Clickhouse support for db migrations with the flyway core API?

nyivanov commented:
Hello @sazonov , I am trying to use your fork for Clickhouse migrations but i cannot get it to work so far. Can you please point me to what am i doing wrong.
My dependencies: 

`org.flyway:flyway-community-db-support,
org.flyway:flyway-core,
com.clickhouse:clickhouse-jdbc,
com.clickhouse:clickhouse-client
`
My migration code: 
`String dbUrl = String.format("jdbc:clickhouse://%s:18123/test_db", dbHost);Flyway flyway = Flyway.configure().locations("filesystem:" + outputFolder).dataSource(dbUrl, dbUsername, dbPassword).load();flyway.migrate();`

I have a simple sql file in the `output` folder: `USE test_db;
SELECT NOW();`

Invoking `flyway.migrate()` i get this exception: `java.lang.NullPointerException: Cannot invoke "String.startsWith(String)" because "identifier" is null` in `ClickHouseConnection::getCurrentSchemaNameOrSearchPath

sazonov commented:
> > No help is needed. PR is waiting to be merged by @Barry-RG. I'm syncing PR with master to keep it actual.
> 
> The Flyway team is working on v10. The new DB support is on hold until the new major version is available (ETA ~ the upcoming fall). @sazonov Could you update the branch once v10 is available? I'll resume the talks with the Flyway team to merge the PR.

PR is now compatible with FlyWay 10. Could you please talk to the team?

sazonov commented:
> PR is now compatible with FlyWay 10. Could you please talk to the team?
@mshustov

mshustov commented:
@sazonov done

herrytco commented:
Is there any estimate on when this will be merged? Waiting for that also ...

sazonov commented:
![image](https://github.com/flyway/flyway/assets/1421938/1b1375be-0007-46fd-873f-77ae7aa16bbe)

uguy commented:
Come on Flyway team, liquibase seems to have no real support for clickhouse (only an old outdated project not even official, but a nice try anyway), could make people choose this tool instead :wink:

ZixuanZhang-96 commented:
@sazonov Appreciate your work! Hope this pr can be merged soon.

arvkonstantin commented:
![image](https://github.com/flyway/flyway/assets/13286752/e48d6b59-bd85-41d6-a786-78d6f5b8497c)

hotstar-xia commented:
Hi team, it's been a while, and v10 is also released. Can we merge it recently? @Mityushin @Hubbitus @DoodleBobBuffPants

johnny commented:
First, thank you @sazonov for keeping this PR alive.

We do have a working private version based on one of the previous iterations, but had some issues getting it to work.

This (as did the previous iterations) works for single node clickhouse clusters, but will not work when using a cluster with multiple shards. I might have missed something, so if the following has any errors, please correct me.

First, the create database statement is missing the "ON CLUSTER" fragment. This will cause the table to only be created on the node the migration runner connects to. Same for the drop database statement.

Second, the migration table needs to be created in a cluster without shards. E.g. the Altinity Operator creates a cluster configuration called 'all-replicated' which adds all nodes of a clickhouse cluster in a single shard.
If the cluster configuration has multiple shards, table rows will only exist in one of these, so migrations will be rerun if the migration runner connects to a different shard.

Third, using a distributed table won't work, since distributed tables are read only and don't allow a rollback/repair of the migrations. Rollback/repair requires deletes to work. Distributed tables were used here https://github.com/flyway/flyway/pull/3333/files#diff-9b0be1a4bf2e38f7ed41c283a5f9bc4fcc1a8c69705aa4398b859d717ab43091R76

johnny commented:
> > This (as did the previous iterations) works for single node clickhouse clusters, but will not work when using a cluster with multiple shards. I might have missed something, so if the following has any errors, please correct me.
> 
> [Replication does not depend on sharding. Each shard has its own independent replication.](https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/replication)
If you only have a single shard, everything works fine. All the replicas in that single shard get all the rows.

> 
> > First, the create database statement is missing the "ON CLUSTER" fragment. This will cause the table to only be created on the node the migration runner connects to. Same for the drop database statement.
> 
> I do see [`ON CLUSTER`](https://github.com/flyway/flyway/blob/0a6c53a527fc1b6f79d256b2d3154162e9fcf311/flyway-community-db-support/flyway-database-clickhouse/src/main/java/org/flywaydb/community/database/clickhouse/ClickHouseDatabase.java#L80) when `clusterName` is configured.
I said "create database" not "create table"
> 
> > Second, the migration table needs to be created in a cluster without shards. E.g. the Altinity Operator creates a cluster configuration called 'all-replicated' which adds all nodes of a clickhouse cluster in a single shard. If the cluster configuration has multiple shards, table rows will only exist in one of these, so migrations will be rerun if the migration runner connects to a different shard.
> 
> Maybe we could be inspired by [this example](https://github.com/golang-migrate/migrate/blob/0d4158977486291b6d148e053a3671574cb01b6c/database/clickhouse/clickhouse.go#L248C62-L248C62) where it is an configuration option?
The cluster configuration name is already configurable here, I think. But the cluster needs to be configured properly (having only one shard in this configuration). This caveat needs to be documented, I think.
> 
> > Third, using a distributed table won't work, since distributed tables are read only and don't allow a rollback/repair of the migrations. Rollback/repair requires deletes to work. Distributed tables were used here https://github.com/flyway/flyway/pull/3333/files#diff-9b0be1a4bf2e38f7ed41c283a5f9bc4fcc1a8c69705aa4398b859d717ab43091R76
> 
> [Searching for `read-only` on the ClickHouse Docs:](https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/replication)
> 
>     1. If ZooKeeper is not set in the config file, you can’t create replicated tables, and any existing replicated tables will be read-only.
> 
>     2. The path in ClickHouse Keeper cannot be changed, and when the table is renamed, the macros will expand into a different path, the table will refer to a path that does not exist in ClickHouse Keeper, and will go into read-only mode.
> 
>     3. If ClickHouse Keeper is unavailable when a server starts, replicated tables switch to read-only mode. The system periodically attempts to connect to ClickHouse Keeper.
> 
> 
> All of these have to do with configuration and only if `clusterName` is configured.
I think you misunderstood. I'm talking about distributed tables, not replicated tables. It's been a while, but it's basically the issue mentioned here: https://stackoverflow.com/questions/76969438/how-to-delete-rows-from-replicateddistributed-tables-on-clickhouse . You can delete the rows in the replicated tables, but not in the distributed tables.
But using the distributed tables would resolve the issue that not all shards get all the rows. Distributed tables are a view across all the shards.

sazonov commented:
> > I think you misunderstood. I'm talking about distributed tables, not replicated tables. It's been a while, but it's basically the issue mentioned here: https://stackoverflow.com/questions/76969438/how-to-delete-rows-from-replicateddistributed-tables-on-clickhouse . You can delete the rows in the replicated tables, but not in the distributed tables.
> > But using the distributed tables would resolve the issue that not all shards get all the rows. Distributed tables are a view across all the shards.
> 
> Now I understand.
> 
> @sazonov Would you accept a pull request (for your pull request) with the engine as a configuration option and using `TinyLog` as the default?

Log engine family does not support mutations (https://clickhouse.com/docs/en/engines/table-engines/log-family), so flyway 'repair' command will not be available. We started with TinyLog as default engine and eventually replaced it with MergeTree that works properly on cluster env.

Could you please provide a docker file to setup your env to reproduce your issue.

johnny commented:
What would be the benefit of using Tinylog? It would not solve any of the
issues above, and as mentioned by Sergey, would be a regression.

A dockerfile would not be enough to reproduce the issues I mentioned. You
need a sharded setup.

Sergey, I'd like to get your opinion on my first message.

On Tue, Jan 9, 2024 at 4:19 PM Yunus ***@***.***> wrote:

> First of all, thank you for your work and the quick reply.
>
> Could you please provide a docker file to setup your env to reproduce your
> issue.
>
> I do not have issues with this branch when using it on a single node or
> clustered.
>
> It seems that other people would prefer to use another engine, by making
> the engine configurable they can decide whether to shoot themselves in the
> foot.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/flyway/flyway/pull/3611#issuecomment-1883247284>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAACIBCE2QA7POLSBCULYDYNVNYRAVCNFSM6AAAAAAUQDXGE2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOBTGI2DOMRYGQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

darkfrog26 commented:
@sazonov Is there any possibility of bringing this small change into this PR: https://github.com/sazonov/flyway/pull/1

Our company needs this functionality.

sazonov commented:
> @sazonov Is there any possibility of bringing this small change into this PR: [sazonov#1](https://github.com/sazonov/flyway/pull/1)
> 
> Our company needs this functionality.

Could you please describe the problem you are solving.

sazonov commented:
> > Could you please describe the problem you are solving.
> 
> Again, instead of asking for someone else's problem. Would you accept a PR that has an option to configure the used engine with the current engines as default if no engine is configured?

It's doesn't work that way. If I'm maintaining something I need to know the purpose of each commit :) Anyway I don't see a link to PR you are mentioned (with engine as configuration param), could you please provide it.

sazonov commented:
> What would be the benefit of using Tinylog? It would not solve any of the issues above, and as mentioned by Sergey, would be a regression. A dockerfile would not be enough to reproduce the issues I mentioned. You need a sharded setup. Sergey, I'd like to get your opinion on my first message.
> […](#)
> On Tue, Jan 9, 2024 at 4:19 PM Yunus ***@***.***> wrote: First of all, thank you for your work and the quick reply. Could you please provide a docker file to setup your env to reproduce your issue. I do not have issues with this branch when using it on a single node or clustered. It seems that other people would prefer to use another engine, by making the engine configurable they can decide whether to shoot themselves in the foot. — Reply to this email directly, view it on GitHub <[#3611 (comment)](https://github.com/flyway/flyway/pull/3611#issuecomment-1883247284)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAACIBCE2QA7POLSBCULYDYNVNYRAVCNFSM6AAAAAAUQDXGE2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOBTGI2DOMRYGQ> . You are receiving this because you commented.Message ID: ***@***.***>

I checked your first message again. It seems you have some specific configuration for ClickHouse cluster. I tested this PR on cluster with 4 nodes and it works well, migration table is created and updated. Could you please provide some more details so I can investigate further.

sazonov commented:
> I checked your first message again. It seems you have some specific configuration for ClickHouse cluster. I tested this PR on cluster with 4 nodes and it works well, migration table is created and updated. Could you please provide some more details so I can investigate further.

This part of my answer was for @johnny.

> I have not worked on it, yet. That is why I am asking if you would be open to it.

@19072376 I'm open to anything that doesn't lead to a problem. Engine type as a configuration param could by implemented via enumerated value which allows you to use the only engine types that don't lead to a regression. This feature needs to be only when you have to have _different_ **proper** behaviours that cannot by achieved with a single engine type. In addition if there is an engine type that is more suitable for most users so it must replace the current one by default. I'm standing for common sense and code simplicity. If your feature just produces some code with no meaning, sorry, not on my sift :)

estaine commented:
> > @sazonov Is there any possibility of bringing this small change into this PR: [sazonov#1](https://github.com/sazonov/flyway/pull/1)
> > Our company needs this functionality.
> 
> Could you please describe the problem you are solving.

Hello, @sazonov, copying my answer from the one in the PR to your repo (answering on behalf of @darkfrog26):

Hey, this PR moves Zookeeper path to a parameter. We use a {uuid} macros-based Zookeeper paths in order not to make an extra dependency between a table name and its path in Zookeper.
The default behaviour (when no config parameter is set) is left intact.

sazonov commented:
> > > @sazonov Is there any possibility of bringing this small change into this PR: [sazonov#1](https://github.com/sazonov/flyway/pull/1)
> > > Our company needs this functionality.
> > 
> > 
> > Could you please describe the problem you are solving.
> 
> Hello, @sazonov, copying my answer from the one in the PR to your repo (answering on behalf of @darkfrog26):
> 
> Hey, this PR moves Zookeeper path to a parameter. We use a {uuid} macros-based Zookeeper paths in order not to make an extra dependency between a table name and its path in Zookeper. The default behaviour (when no config parameter is set) is left intact.

Merged.

johnny commented:
Steps to reproduce the issue with the missing "create database on cluster"
1. A cluster with at least 2 shards
2. Execute the migration against the cluster which basically does
> CREATE DATABASE 'foo'
> CREATE TABLE IF NOT EXISTS 'foo.table' ON CLUSTER '{cluster}'
3. Migration fails, since the database only exists on one shard

Steps to reproduce the issue with the replicated merge tree not spanning
the whole cluster:
1. A cluster with at least 2 shards
2. Execute migration against the cluster (this time with CREATE DATABASE ON
CLUSTER '{cluster}'
3. Connect to one shard explicitly and list contents of the flyway
migrations table
4. Repeat 3 until you find an empty table

On Thu, Jan 11, 2024 at 8:58 AM Sergey Sazonov ***@***.***>
wrote:

> I checked your first message again. It seems you have some specific
> configuration for ClickHouse cluster. I tested this PR on cluster with 4
> nodes and it works well, migration table is created and updated. Could you
> please provide some more details so I can investigate further.
>
> This part of my answer was for @johnny <https://github.com/johnny>.
>
> I have not worked on it, yet. That is why I am asking if you would be open
> to it.
>
> @19072376 <https://github.com/19072376> I'm open to anything that doesn't
> lead to a problem. Engine type as a configuration param could by
> implemented via enumerated value which allows you to use the only engine
> types that don't lead to a regression. This feature needs to be only when
> you have to have *different* *proper* behaviours that cannot by achieved
> with a single engine type. In addition if there is an engine type that is
> more suitable for most users so it must replace the current one by default.
> I'm standing for common sense and code simplicity. If your feature just
> produces some code with no meaning, sorry, not on my sift :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/flyway/flyway/pull/3611#issuecomment-1886562246>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAACIHWR3BG2TNHYXCOYMDYN6LSDAVCNFSM6AAAAAAUQDXGE2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOBWGU3DEMRUGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

good-vi commented:
What are flyway team waiting for?)

darkfrog26 commented:
@sazonov I'm happy to assist if there's anything I can do to help.

Barry-RG commented on flyway-community-db-support/flyway-database-clickhouse/src/main/java/org/flywaydb/community/database/clickhouse/ClickHouseConfigurationExtension.java:
```diff
@@ -0,0 +1,61 @@
+@Getter
+public class ClickHouseConfigurationExtension implements ConfigurationExtension {
+    private static final String CLUSTER_NAME = "flyway.clickhouse.clusterName";
+    private static final String ZOOKEEPER_PATH = "flyway.clickhouse.zookeperPath";
```
@sazonov I have noticed what I think is a typo, can this be confirmed?
Is `flyway.clickhouse.zookeperPath` meant to be `flyway.clickhouse.zookeeperPath`?

sazonov replied:
Confirmed, that's a typo. Fixed that.

Barry-RG replied:
Thanks @sazonov That was going to be a follow up as I saw that not being used.

Barry-RG reviewed (changes requested):
Noticed what I think is a typo

Barry-RG commented:
This has now been release from https://github.com/flyway/flyway-community-db-support

KingSeybro commented:
> Steps to reproduce the issue with the missing "create database on cluster" 1. A cluster with at least 2 shards 2. Execute the migration against the cluster which basically does
> CREATE DATABASE 'foo' CREATE TABLE IF NOT EXISTS 'foo.table' ON CLUSTER '{cluster}'
> 3. Migration fails, since the database only exists on one shard Steps to reproduce the issue with the replicated merge tree not spanning the whole cluster: 1. A cluster with at least 2 shards 2. Execute migration against the cluster (this time with CREATE DATABASE ON CLUSTER '{cluster}' 3. Connect to one shard explicitly and list contents of the flyway migrations table 4. Repeat 3 until you find an empty table
> […](#)
> On Thu, Jan 11, 2024 at 8:58 AM Sergey Sazonov ***@***.***> wrote: I checked your first message again. It seems you have some specific configuration for ClickHouse cluster. I tested this PR on cluster with 4 nodes and it works well, migration table is created and updated. Could you please provide some more details so I can investigate further. This part of my answer was for @johnny <https://github.com/johnny>. I have not worked on it, yet. That is why I am asking if you would be open to it. @19072376 <https://github.com/19072376> I'm open to anything that doesn't lead to a problem. Engine type as a configuration param could by implemented via enumerated value which allows you to use the only engine types that don't lead to a regression. This feature needs to be only when you have to have *different* *proper* behaviours that cannot by achieved with a single engine type. In addition if there is an engine type that is more suitable for most users so it must replace the current one by default. I'm standing for common sense and code simplicity. If your feature just produces some code with no meaning, sorry, not on my sift :) — Reply to this email directly, view it on GitHub <[#3611 (comment)](https://github.com/flyway/flyway/pull/3611#issuecomment-1886562246)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAACIHWR3BG2TNHYXCOYMDYN6LSDAVCNFSM6AAAAAAUQDXGE2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOBWGU3DEMRUGY> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Did you by any chance solve this issue? We have something (maybe) similar. 2 shard setup, Spring Boot Application. Our flyway_history_table gets created as a MergeTree instead of a ReplicatedMergeTree. I tried setting the environment variable `FLYWAY_CLICKHOUSE_CLUSTER_NAME` as well as the configuration properties `flyway.clickhouse.clusterName`, but nothing seems to work
